coreppl/coreppl.mc

```diff
@@ -10,76 +10,191 @@ include "mexpr/ast-builder.mc"
+    match lhs with TmInfer l then
+      eqExprH env free l.model r.model
+    else None ()
+
 end
 
 
 -- Assume defines a new random variable
-lang Assume = Ast + Dist + PrettyPrint
+lang Assume = Ast + Dist + PrettyPrint + Eq
+
   syn Expr =
   | TmAssume { dist: Expr,
-               info: Info}
+               ty: Type,
+               info: Info }
+
+  sem infoTm =
+  | TmAssume t -> t.info
+
+  sem ty =
+  | TmAssume t -> t.ty
+
+  sem withType (ty: Type) =
+  | TmAssume t -> TmAssume { t with ty = ty }
 
+  sem smap_Expr_Expr (f: Expr -> a) =
+  | TmAssume t -> TmAssume { t with dist = f t.dist }
+
+  sem sfold_Expr_Expr (f: a -> b -> a) (acc: a) =
+  | TmAssume t -> f acc t.dist
+
+  -- Pretty printing
   sem isAtomic =
   | TmAssume _ -> false
 
   sem pprintCode (indent : Int) (env: PprintEnv) =
-  | TmAssume r ->
-    match pprintCode 0 env r.dist with (env,dist) then
-      (env, join ["assume ", dist])
+  | TmAssume t ->
+    let i = pprintIncr indent in
+    match printParen i env t.dist with (env,dist) then
+      (env, join ["assume", pprintNewline i, dist])
     else never
 
+  -- Equality
+  sem eqExprH (env : EqEnv) (free : EqEnv) (lhs : Expr) =
+  | TmAssume r ->
+    match lhs with TmAssume l then
+      eqExprH env free l.dist r.dist
+    else None ()
+
 end
 
 
 -- Observe gives a random variable conditioned on a specific value
-lang Observe = Ast + Dist + PrettyPrint
+lang Observe = Ast + Dist + PrettyPrint + Eq
+
   syn Expr =
   | TmObserve { value: Expr,
                 dist: Expr,
+                ty: Type,
                 info: Info }
 
+  sem infoTm =
+  | TmObserve t -> t.info
+
+  sem ty =
+  | TmObserve t -> t.ty
+
+  sem withType (ty: Type) =
+  | TmObserve t -> TmObserve { t with ty = ty }
+
+  sem smap_Expr_Expr (f: Expr -> a) =
+  | TmObserve t -> TmObserve {{ t with value = f t.value }
+                                  with dist = f t.dist }
+
+  sem sfold_Expr_Expr (f: a -> b -> a) (acc: a) =
+  | TmObserve t -> f (f acc t.value) t.dist
+
+  -- Pretty printing
   sem isAtomic =
   | TmObserve _ -> false
 
   sem pprintCode (indent : Int) (env: PprintEnv) =
-  | TmObserve r ->
-    match pprintCode 0 env r.value with (env, value) then
-      match pprintCode 0 env r.dist with (env, dist) then
-        (env, join ["observe ", value, " ", dist])
-      else never
+  | TmObserve t ->
+    let i = pprintIncr indent in
+    match printArgs i env [t.value, t.dist] with (env,args) then
+      (env, join ["observe", pprintNewline i, args])
     else never
 
+  -- Equality
+  sem eqExprH (env : EqEnv) (free : EqEnv) (lhs : Expr) =
+  | TmObserve r ->
+    match lhs with TmObserve l then
+      match eqExprH env free l.value r.value with Some free then
+        eqExprH env free l.dist r.dist
+      else None ()
+    else None ()
+
 end
 
 -- Defines a weight term
-lang Weight = Ast + PrettyPrint
+lang Weight = Ast + PrettyPrint + Eq
   syn Expr =
   | TmWeight { weight: Expr }
```

Comment:
I guess `info` and `ty` fields should be added

Reply:
Yes, thanks. I'll fix it